diff --git a/dist/ModuleFederationPlugin.js b/dist/ModuleFederationPlugin.js
index 2818028..c20ab1f 100644
--- a/dist/ModuleFederationPlugin.js
+++ b/dist/ModuleFederationPlugin.js
@@ -69,8 +69,6 @@ const sdk_namespaceObject = require("@module-federation/sdk");
 const manifest_namespaceObject = require("@module-federation/manifest");
 ;// CONCATENATED MODULE: external "@module-federation/managers"
 const managers_namespaceObject = require("@module-federation/managers");
-;// CONCATENATED MODULE: external "@module-federation/dts-plugin"
-const dts_plugin_namespaceObject = require("@module-federation/dts-plugin");
 ;// CONCATENATED MODULE: external "@module-federation/bridge-react-webpack-plugin"
 const bridge_react_webpack_plugin_namespaceObject = require("@module-federation/bridge-react-webpack-plugin");
 var bridge_react_webpack_plugin_default = /*#__PURE__*/__webpack_require__.n(bridge_react_webpack_plugin_namespaceObject);
@@ -201,7 +199,8 @@ class ModuleFederationPlugin {
         let disableManifest = options.manifest === false;
         let disableDts = options.dts === false;
         if (!disableDts) {
-            const dtsPlugin = new dts_plugin_namespaceObject.DtsPlugin(options);
+            const { DtsPlugin } = require("@module-federation/dts-plugin");
+            const dtsPlugin = new DtsPlugin(options);
             // @ts-ignore
             dtsPlugin.apply(compiler);
             dtsPlugin.addRuntimePlugins();
diff --git a/dist/ModuleFederationPlugin.mjs b/dist/ModuleFederationPlugin.mjs
index 2daf1a6..65b2c88 100644
--- a/dist/ModuleFederationPlugin.mjs
+++ b/dist/ModuleFederationPlugin.mjs
@@ -1,12 +1,13 @@
 import { bindLoggerToCompiler, composeKeyWithSeparator } from "@module-federation/sdk";
 import { StatsPlugin } from "@module-federation/manifest";
 import { ContainerManager, utils } from "@module-federation/managers";
-import { DtsPlugin } from "@module-federation/dts-plugin";
+import { createRequire } from "node:module";
 import bridge_react_webpack_plugin from "@module-federation/bridge-react-webpack-plugin";
 import node_path from "node:path";
 import node_fs from "node:fs";
 import { RemoteEntryPlugin } from "./RemoteEntryPlugin.mjs";
 import logger from "./logger.mjs";
+const require = createRequire(import.meta.url);
 
 
 
@@ -131,6 +132,7 @@ class ModuleFederationPlugin {
         let disableManifest = options.manifest === false;
         let disableDts = options.dts === false;
         if (!disableDts) {
+            const { DtsPlugin } = require("@module-federation/dts-plugin");
             const dtsPlugin = new DtsPlugin(options);
             // @ts-ignore
             dtsPlugin.apply(compiler);
