diff --git a/dist/utils.js b/dist/utils.js
--- a/dist/utils.js
+++ b/dist/utils.js
@@ -71,8 +71,6 @@ const external_path_namespaceObject = require("path");
 var external_path_default = /*#__PURE__*/__webpack_require__.n(external_path_namespaceObject);
 ;// CONCATENATED MODULE: external "@module-federation/sdk"
 const sdk_namespaceObject = require("@module-federation/sdk");
-;// CONCATENATED MODULE: external "@module-federation/dts-plugin/core"
-const core_namespaceObject = require("@module-federation/dts-plugin/core");
 ;// CONCATENATED MODULE: external "./constants.js"
 const external_constants_js_namespaceObject = require("./constants.js");
 ;// CONCATENATED MODULE: external "./logger.js"
@@ -252,7 +250,11 @@ function getTypesMetaInfo(pluginOptions, context) {
         api: ''
     };
     try {
-        const normalizedDtsOptions = (0,sdk_namespaceObject.normalizeOptions)((0,core_namespaceObject.isTSProject)(pluginOptions.dts, context), {
+        if (pluginOptions.dts === false) {
+            return defaultTypesMetaInfo;
+        }
+        const { isTSProject, retrieveTypesAssetsInfo } = require("@module-federation/dts-plugin/core");
+        const normalizedDtsOptions = (0,sdk_namespaceObject.normalizeOptions)(isTSProject(pluginOptions.dts, context), {
             generateTypes: defaultRemoteOptions,
             consumeTypes: {}
         }, 'mfOptions.dts')(pluginOptions.dts);
@@ -263,7 +265,7 @@ function getTypesMetaInfo(pluginOptions, context) {
         if (normalizedRemote === false) {
             return defaultTypesMetaInfo;
         }
-        const { apiFileName, zipName } = (0,core_namespaceObject.retrieveTypesAssetsInfo)({
+        const { apiFileName, zipName } = retrieveTypesAssetsInfo({
             ...normalizedRemote,
             context,
             moduleFederationConfig: pluginOptions
diff --git a/dist/utils.mjs b/dist/utils.mjs
--- a/dist/utils.mjs
+++ b/dist/utils.mjs
@@ -1,8 +1,9 @@
 import path from "path";
 import { normalizeOptions } from "@module-federation/sdk";
-import { isTSProject, retrieveTypesAssetsInfo } from "@module-federation/dts-plugin/core";
+import { createRequire } from "node:module";
 import { HOT_UPDATE_SUFFIX, PLUGIN_IDENTIFIER } from "./constants.mjs";
 import logger from "./logger.mjs";
+const require = createRequire(import.meta.url);
 
 
 
@@ -182,6 +183,10 @@ function getTypesMetaInfo(pluginOptions, context) {
         api: ''
     };
     try {
+        if (pluginOptions.dts === false) {
+            return defaultTypesMetaInfo;
+        }
+        const { isTSProject, retrieveTypesAssetsInfo } = require("@module-federation/dts-plugin/core");
         const normalizedDtsOptions = normalizeOptions(isTSProject(pluginOptions.dts, context), {
             generateTypes: defaultRemoteOptions,
             consumeTypes: {}
