diff --git a/dist/esm/fork-dev-worker.js b/dist/esm/fork-dev-worker.js
index d21d51e..fe887a2 100644
--- a/dist/esm/fork-dev-worker.js
+++ b/dist/esm/fork-dev-worker.js
@@ -135,7 +135,7 @@ process.on("message", (message) => {
   fileLog(`ChildProcess(${process.pid}), message: ${JSON.stringify(message)} `, "forkDevWorker", "info");
   if (message.type === rpc_exports.RpcGMCallTypes.EXIT) {
     fileLog(`ChildProcess(${process.pid}) SIGTERM, Federation DevServer will exit...`, "forkDevWorker", "error");
-    moduleServer.exit();
+    moduleServer?.exit();
     process.exit(0);
   }
 });
diff --git a/dist/fork-dev-worker.js b/dist/fork-dev-worker.js
index 94b2e8d..399b9ba 100644
--- a/dist/fork-dev-worker.js
+++ b/dist/fork-dev-worker.js
@@ -2862,7 +2862,7 @@ process.on("message", (message) => {
   fileLog(`ChildProcess(${process.pid}), message: ${JSON.stringify(message)} `, "forkDevWorker", "info");
   if (message.type === rpc_exports.RpcGMCallTypes.EXIT) {
     fileLog(`ChildProcess(${process.pid}) SIGTERM, Federation DevServer will exit...`, "forkDevWorker", "error");
-    moduleServer.exit();
+    moduleServer?.exit();
     process.exit(0);
   }
 });
