diff --git a/cjs/react-server-dom-rspack-client.browser.development.js b/cjs/react-server-dom-rspack-client.browser.development.js
index 950a74c..2aecc89 100644
--- a/cjs/react-server-dom-rspack-client.browser.development.js
+++ b/cjs/react-server-dom-rspack-client.browser.development.js
@@ -555,7 +555,7 @@
             null === formData && (formData = new FormData());
             var _data3 = formData;
             key = nextPartId++;
-            var prefix = formFieldPrefix + key + "_";
+            var prefix = formFieldPrefix + "_" + key + "_";
             value.forEach(function (originalValue, originalKey) {
               _data3.append(prefix + originalKey, originalValue);
             });
diff --git a/cjs/react-server-dom-rspack-client.browser.production.js b/cjs/react-server-dom-rspack-client.browser.production.js
index 5ebfbf8..4bf2908 100644
--- a/cjs/react-server-dom-rspack-client.browser.production.js
+++ b/cjs/react-server-dom-rspack-client.browser.production.js
@@ -359,7 +359,7 @@ function processReply(
         null === formData && (formData = new FormData());
         var data$31 = formData;
         key = nextPartId++;
-        var prefix = formFieldPrefix + key + "_";
+        var prefix = formFieldPrefix + "_" + key + "_";
         value.forEach(function (originalValue, originalKey) {
           data$31.append(prefix + originalKey, originalValue);
         });
diff --git a/cjs/react-server-dom-rspack-client.edge.development.js b/cjs/react-server-dom-rspack-client.edge.development.js
index 8e18c2e..1d31257 100644
--- a/cjs/react-server-dom-rspack-client.edge.development.js
+++ b/cjs/react-server-dom-rspack-client.edge.development.js
@@ -576,7 +576,7 @@
             null === formData && (formData = new FormData());
             var _data3 = formData;
             key = nextPartId++;
-            var prefix = formFieldPrefix + key + "_";
+            var prefix = formFieldPrefix + "_" + key + "_";
             value.forEach(function (originalValue, originalKey) {
               _data3.append(prefix + originalKey, originalValue);
             });
diff --git a/cjs/react-server-dom-rspack-client.edge.production.js b/cjs/react-server-dom-rspack-client.edge.production.js
index deba87e..e6fba77 100644
--- a/cjs/react-server-dom-rspack-client.edge.production.js
+++ b/cjs/react-server-dom-rspack-client.edge.production.js
@@ -358,7 +358,7 @@ function processReply(
         null === formData && (formData = new FormData());
         var data$31 = formData;
         key = nextPartId++;
-        var prefix = formFieldPrefix + key + "_";
+        var prefix = formFieldPrefix + "_" + key + "_";
         value.forEach(function (originalValue, originalKey) {
           data$31.append(prefix + originalKey, originalValue);
         });
diff --git a/cjs/react-server-dom-rspack-client.node.development.js b/cjs/react-server-dom-rspack-client.node.development.js
index 83bf6e8..9b8a174 100644
--- a/cjs/react-server-dom-rspack-client.node.development.js
+++ b/cjs/react-server-dom-rspack-client.node.development.js
@@ -576,7 +576,7 @@
             null === formData && (formData = new FormData());
             var _data3 = formData;
             key = nextPartId++;
-            var prefix = formFieldPrefix + key + "_";
+            var prefix = formFieldPrefix + "_" + key + "_";
             value.forEach(function (originalValue, originalKey) {
               _data3.append(prefix + originalKey, originalValue);
             });
diff --git a/cjs/react-server-dom-rspack-client.node.production.js b/cjs/react-server-dom-rspack-client.node.production.js
index a23b667..f422d2a 100644
--- a/cjs/react-server-dom-rspack-client.node.production.js
+++ b/cjs/react-server-dom-rspack-client.node.production.js
@@ -359,7 +359,7 @@ function processReply(
         null === formData && (formData = new FormData());
         var data$31 = formData;
         key = nextPartId++;
-        var prefix = formFieldPrefix + key + "_";
+        var prefix = formFieldPrefix + "_" + key + "_";
         value.forEach(function (originalValue, originalKey) {
           data$31.append(prefix + originalKey, originalValue);
         });
diff --git a/cjs/react-server-dom-rspack-server.browser.development.js b/cjs/react-server-dom-rspack-server.browser.development.js
index 3534775..515c3c5 100644
--- a/cjs/react-server-dom-rspack-server.browser.development.js
+++ b/cjs/react-server-dom-rspack-server.browser.development.js
@@ -1331,7 +1331,7 @@
           erroredTask(request, streamTask, reason),
           enqueueFlush(request),
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error));
+            iterator.throw(reason).then(noop, noop));
       }
       function abortIterable() {
         if (0 === streamTask.status) {
@@ -1345,7 +1345,7 @@
             : (erroredTask(request, streamTask, signal.reason),
               enqueueFlush(request));
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error);
+            iterator.throw(reason).then(noop, noop);
         }
       }
       var isIterator = iterable === iterator,
@@ -3834,6 +3834,22 @@
       chunkMap.set(chunkId, filename);
       return __webpack_chunk_load__(chunkId);
     }
+    function appendBackingEntry(backingStore, key, value) {
+      backingStore.data.append(key, value);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
+    function appendBackingFile(backingStore, key, value, filename) {
+      backingStore.data.append(key, value, filename);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
     function ReactPromise(status, value, reason) {
       this.status = status;
       this.value = value;
@@ -4146,7 +4162,7 @@
       var chunks = response._chunks,
         chunk = chunks.get(id);
       chunk ||
-        ((chunk = response._formData.get(response._prefix + id)),
+        ((chunk = response._formData.data.get(response._prefix + id)),
         (chunk =
           "string" === typeof chunk
             ? new ReactPromise(
@@ -4258,6 +4274,10 @@
         case "fulfilled":
           id = chunk.value;
           chunk = chunk.reason;
+          if (null !== chunk && "error" in chunk)
+            throw Error(
+              "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+            );
           for (
             var localLength = 0,
               rootArrayContexts = response._rootArrayContexts,
@@ -4349,23 +4369,20 @@
     function createMap(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Map initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Map.");
-      response = new Map(model);
       model.$$consumed = !0;
-      return response;
+      return new Map(model);
     }
     function createSet(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Set initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Set.");
-      response = new Set(model);
       model.$$consumed = !0;
-      return response;
+      return new Set(model);
     }
     function extractIterator(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Iterator initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Iterator.");
-      response = model[Symbol.iterator]();
       model.$$consumed = !0;
-      return response;
+      return model[Symbol.iterator]();
     }
     function createModel(response, model, parentObject, key) {
       return "then" === key && "function" === typeof model ? null : model;
@@ -4403,7 +4420,7 @@
           Error("Already initialized typed array.")
         )
       );
-      reference = response._formData.get(key).arrayBuffer();
+      reference = response._formData.data.get(key).arrayBuffer();
       if (initializingHandler) {
         var handler = initializingHandler;
         handler.deps++;
@@ -4447,7 +4464,7 @@
       var chunks = response._chunks;
       stream = new ReactPromise("fulfilled", stream, controller);
       chunks.set(id, stream);
-      response = response._formData.getAll(response._prefix + id);
+      response = response._formData.data.getAll(response._prefix + id);
       for (id = 0; id < response.length; id++)
         (chunks = response[id]),
           "string" === typeof chunks &&
@@ -4667,24 +4684,33 @@
               getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
             );
           case "K":
-            obj = value.slice(2);
-            obj = response._prefix + obj + "_";
-            key = new FormData();
-            response = response._formData;
-            arrayRoot = Array.from(response.keys());
-            for (value = 0; value < arrayRoot.length; value++)
-              if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+            key = value.slice(2);
+            obj = response._prefix + "_";
+            key = obj + key + "_";
+            arrayRoot = new FormData();
+            for (response = response._formData; ; ) {
+              value = response;
+              reference = value.keys;
+              null === reference &&
+                ((reference = value.keys = Array.from(value.data.keys())),
+                (value.keyPointer = 0));
+              value = reference[value.keyPointer];
+              if (void 0 === value) break;
+              if (value.startsWith(key)) {
+                reference = response.data.getAll(value);
                 for (
-                  var entries = response.getAll(reference),
-                    newKey = reference.slice(obj.length),
-                    j = 0;
-                  j < entries.length;
-                  j++
+                  var referencedFormDataKey = value.slice(key.length), i = 0;
+                  i < reference.length;
+                  i++
                 )
-                  key.append(newKey, entries[j]);
-                response.delete(reference);
-              }
-            return key;
+                  arrayRoot.append(referencedFormDataKey, reference[i]);
+                reference = response;
+                reference.data.delete(value);
+                reference.keyPointer++;
+              } else if (value.startsWith(obj)) break;
+              else response.keyPointer++;
+            }
+            return arrayRoot;
           case "i":
             return (
               (arrayRoot = value.slice(2)),
@@ -4851,10 +4877,11 @@
               arrayRoot
             );
           case "B":
-            return (
-              (obj = parseInt(value.slice(2), 16)),
-              response._formData.get(response._prefix + obj)
-            );
+            obj = parseInt(value.slice(2), 16);
+            response = response._formData.data.get(response._prefix + obj);
+            if (!(response instanceof Blob))
+              throw Error("Referenced Blob is not a Blob.");
+            return response;
           case "R":
             return parseReadableStream(response, value, void 0);
           case "r":
@@ -4892,7 +4919,7 @@
       return {
         _bundlerConfig: bundlerConfig,
         _prefix: formFieldPrefix,
-        _formData: backingFormData,
+        _formData: { data: backingFormData, keyPointer: -1, keys: null },
         _chunks: chunks,
         _closed: !1,
         _closedReason: null,
@@ -4951,44 +4978,44 @@
     }
     function decodeAction$1(body, serverManifest) {
       var formData = new FormData(),
-        action = null,
-        seenActions = new Set();
+        maybeActionKey = null;
       body.forEach(function (value, key) {
         key.startsWith("$ACTION_")
           ? key.startsWith("$ACTION_REF_")
-            ? seenActions.has(key) ||
-              (seenActions.add(key),
-              (value = "$ACTION_" + key.slice(12) + ":"),
-              (value = decodeBoundActionMetaData(body, serverManifest, value)),
-              (action = loadServerReference(serverManifest, value)))
-            : key.startsWith("$ACTION_ID_") &&
-              !seenActions.has(key) &&
-              (seenActions.add(key),
-              (value = key.slice(11)),
-              (action = loadServerReference(serverManifest, {
-                id: value,
-                bound: null
-              })))
+            ? (maybeActionKey = key)
+            : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
           : formData.append(key, value);
       });
-      return null === action
-        ? null
-        : action.then(function (fn) {
-            return fn.bind(null, formData);
-          });
+      if (null === maybeActionKey) return null;
+      var actionKey = maybeActionKey,
+        action = null;
+      if (actionKey.startsWith("$ACTION_REF_"))
+        (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+          (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+          (action = loadServerReference(serverManifest, body));
+      else if (actionKey.startsWith("$ACTION_ID_"))
+        (body = actionKey.slice(11)),
+          (action = loadServerReference(serverManifest, {
+            id: body,
+            bound: null
+          }));
+      else throw Error("Cannot handle action key. This is a bug in React.");
+      return action.then(function (fn) {
+        return fn.bind(null, formData);
+      });
     }
     function decodeFormState$1(actionResult, body, serverManifest) {
       var keyPath = body.get("$ACTION_KEY");
       if ("string" !== typeof keyPath) return Promise.resolve(null);
-      var metaData = null;
+      var actionKey = null;
       body.forEach(function (value, key) {
-        key.startsWith("$ACTION_REF_") &&
-          ((value = "$ACTION_" + key.slice(12) + ":"),
-          (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+        key.startsWith("$ACTION_REF_") && (actionKey = key);
       });
-      if (null === metaData) return Promise.resolve(null);
-      var referenceId = metaData.id;
-      return Promise.resolve(metaData.bound).then(function (bound) {
+      if (null === actionKey) return Promise.resolve(null);
+      var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+      body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+      var referenceId = body.id;
+      return Promise.resolve(body.bound).then(function (bound) {
         return null === bound
           ? null
           : [actionResult, keyPath, referenceId, bound.length - 1];
diff --git a/cjs/react-server-dom-rspack-server.browser.production.js b/cjs/react-server-dom-rspack-server.browser.production.js
index 1af2338..597e787 100644
--- a/cjs/react-server-dom-rspack-server.browser.production.js
+++ b/cjs/react-server-dom-rspack-server.browser.production.js
@@ -931,7 +931,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
       erroredTask(request, streamTask, reason),
       enqueueFlush(request),
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error));
+        iterator.throw(reason).then(noop, noop));
   }
   function abortIterable() {
     if (0 === streamTask.status) {
@@ -945,7 +945,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
         : (erroredTask(request, streamTask, signal.reason),
           enqueueFlush(request));
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error);
+        iterator.throw(reason).then(noop, noop);
     }
   }
   iterable = iterable === iterator;
@@ -2138,6 +2138,22 @@ __webpack_require__.u = function (chunkId) {
   return void 0 !== flightChunk ? flightChunk : rspackGetChunkFilename(chunkId);
 };
 var RESPONSE_SYMBOL = Symbol();
+function appendBackingEntry(backingStore, key, value) {
+  backingStore.data.append(key, value);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
+function appendBackingFile(backingStore, key, value, filename) {
+  backingStore.data.append(key, value, filename);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
 function ReactPromise(status, value, reason) {
   this.status = status;
   this.value = value;
@@ -2504,7 +2520,7 @@ function getChunk(response, id) {
   var chunks = response._chunks,
     chunk = chunks.get(id);
   chunk ||
-    ((chunk = response._formData.get(response._prefix + id)),
+    ((chunk = response._formData.data.get(response._prefix + id)),
     (chunk =
       "string" === typeof chunk
         ? createResolvedModelChunk(response, chunk, id)
@@ -2605,6 +2621,10 @@ function getOutlinedModel(
     case "fulfilled":
       id = chunk.value;
       chunk = chunk.reason;
+      if (null !== chunk && "error" in chunk)
+        throw Error(
+          "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+        );
       for (
         var localLength = 0,
           rootArrayContexts = response._rootArrayContexts,
@@ -2687,23 +2707,20 @@ function getOutlinedModel(
 function createMap(response, model) {
   if (!isArrayImpl(model)) throw Error(formatProdErrorMessage(575));
   if (!0 === model.$$consumed) throw Error(formatProdErrorMessage(572));
-  response = new Map(model);
   model.$$consumed = !0;
-  return response;
+  return new Map(model);
 }
 function createSet(response, model) {
   if (!isArrayImpl(model)) throw Error(formatProdErrorMessage(576));
   if (!0 === model.$$consumed) throw Error(formatProdErrorMessage(573));
-  response = new Set(model);
   model.$$consumed = !0;
-  return response;
+  return new Set(model);
 }
 function extractIterator(response, model) {
   if (!isArrayImpl(model)) throw Error(formatProdErrorMessage(577));
   if (!0 === model.$$consumed) throw Error(formatProdErrorMessage(578));
-  response = model[Symbol.iterator]();
   model.$$consumed = !0;
-  return response;
+  return model[Symbol.iterator]();
 }
 function createModel(response, model, parentObject, key) {
   return "then" === key && "function" === typeof model ? null : model;
@@ -2736,7 +2753,7 @@ function parseTypedArray(
     reference,
     createErroredChunk(response, Error(formatProdErrorMessage(568)))
   );
-  reference = response._formData.get(key).arrayBuffer();
+  reference = response._formData.data.get(key).arrayBuffer();
   if (initializingHandler) {
     var handler = initializingHandler;
     handler.deps++;
@@ -2780,7 +2797,7 @@ function resolveStream(response, id, stream, controller) {
   var chunks = response._chunks;
   stream = new ReactPromise("fulfilled", stream, controller);
   chunks.set(id, stream);
-  response = response._formData.getAll(response._prefix + id);
+  response = response._formData.data.getAll(response._prefix + id);
   for (id = 0; id < response.length; id++)
     (chunks = response[id]),
       "string" === typeof chunks &&
@@ -2988,24 +3005,33 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
         );
       case "K":
-        obj = value.slice(2);
-        obj = response._prefix + obj + "_";
-        key = new FormData();
-        response = response._formData;
-        arrayRoot = Array.from(response.keys());
-        for (value = 0; value < arrayRoot.length; value++)
-          if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+        key = value.slice(2);
+        obj = response._prefix + "_";
+        key = obj + key + "_";
+        arrayRoot = new FormData();
+        for (response = response._formData; ; ) {
+          value = response;
+          reference = value.keys;
+          null === reference &&
+            ((reference = value.keys = Array.from(value.data.keys())),
+            (value.keyPointer = 0));
+          value = reference[value.keyPointer];
+          if (void 0 === value) break;
+          if (value.startsWith(key)) {
+            reference = response.data.getAll(value);
             for (
-              var entries = response.getAll(reference),
-                newKey = reference.slice(obj.length),
-                j = 0;
-              j < entries.length;
-              j++
+              var referencedFormDataKey = value.slice(key.length), i = 0;
+              i < reference.length;
+              i++
             )
-              key.append(newKey, entries[j]);
-            response.delete(reference);
-          }
-        return key;
+              arrayRoot.append(referencedFormDataKey, reference[i]);
+            reference = response;
+            reference.data.delete(value);
+            reference.keyPointer++;
+          } else if (value.startsWith(obj)) break;
+          else response.keyPointer++;
+        }
+        return arrayRoot;
       case "i":
         return (
           (arrayRoot = value.slice(2)),
@@ -3158,10 +3184,11 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           arrayRoot
         );
       case "B":
-        return (
-          (obj = parseInt(value.slice(2), 16)),
-          response._formData.get(response._prefix + obj)
-        );
+        obj = parseInt(value.slice(2), 16);
+        response = response._formData.data.get(response._prefix + obj);
+        if (!(response instanceof Blob))
+          throw Error("Referenced Blob is not a Blob.");
+        return response;
       case "R":
         return parseReadableStream(response, value, void 0);
       case "r":
@@ -3188,7 +3215,7 @@ function createResponse(bundlerConfig, formFieldPrefix, temporaryReferences) {
   return {
     _bundlerConfig: bundlerConfig,
     _prefix: formFieldPrefix,
-    _formData: backingFormData,
+    _formData: { data: backingFormData, keyPointer: -1, keys: null },
     _chunks: chunks,
     _closed: !1,
     _closedReason: null,
@@ -3238,44 +3265,44 @@ function decodeBoundActionMetaData(
 }
 function decodeAction$1(body, serverManifest) {
   var formData = new FormData(),
-    action = null,
-    seenActions = new Set();
+    maybeActionKey = null;
   body.forEach(function (value, key) {
     key.startsWith("$ACTION_")
       ? key.startsWith("$ACTION_REF_")
-        ? seenActions.has(key) ||
-          (seenActions.add(key),
-          (value = "$ACTION_" + key.slice(12) + ":"),
-          (value = decodeBoundActionMetaData(body, serverManifest, value)),
-          (action = loadServerReference(serverManifest, value)))
-        : key.startsWith("$ACTION_ID_") &&
-          !seenActions.has(key) &&
-          (seenActions.add(key),
-          (value = key.slice(11)),
-          (action = loadServerReference(serverManifest, {
-            id: value,
-            bound: null
-          })))
+        ? (maybeActionKey = key)
+        : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
       : formData.append(key, value);
   });
-  return null === action
-    ? null
-    : action.then(function (fn) {
-        return fn.bind(null, formData);
-      });
+  if (null === maybeActionKey) return null;
+  var actionKey = maybeActionKey,
+    action = null;
+  if (actionKey.startsWith("$ACTION_REF_"))
+    (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+      (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+      (action = loadServerReference(serverManifest, body));
+  else if (actionKey.startsWith("$ACTION_ID_"))
+    (body = actionKey.slice(11)),
+      (action = loadServerReference(serverManifest, {
+        id: body,
+        bound: null
+      }));
+  else throw Error("Cannot handle action key. This is a bug in React.");
+  return action.then(function (fn) {
+    return fn.bind(null, formData);
+  });
 }
 function decodeFormState$1(actionResult, body, serverManifest) {
   var keyPath = body.get("$ACTION_KEY");
   if ("string" !== typeof keyPath) return Promise.resolve(null);
-  var metaData = null;
+  var actionKey = null;
   body.forEach(function (value, key) {
-    key.startsWith("$ACTION_REF_") &&
-      ((value = "$ACTION_" + key.slice(12) + ":"),
-      (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+    key.startsWith("$ACTION_REF_") && (actionKey = key);
   });
-  if (null === metaData) return Promise.resolve(null);
-  var referenceId = metaData.id;
-  return Promise.resolve(metaData.bound).then(function (bound) {
+  if (null === actionKey) return Promise.resolve(null);
+  var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+  body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+  var referenceId = body.id;
+  return Promise.resolve(body.bound).then(function (bound) {
     return null === bound
       ? null
       : [actionResult, keyPath, referenceId, bound.length - 1];
diff --git a/cjs/react-server-dom-rspack-server.edge.development.js b/cjs/react-server-dom-rspack-server.edge.development.js
index cd06f82..8db3f26 100644
--- a/cjs/react-server-dom-rspack-server.edge.development.js
+++ b/cjs/react-server-dom-rspack-server.edge.development.js
@@ -1343,7 +1343,7 @@
           erroredTask(request, streamTask, reason),
           enqueueFlush(request),
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error));
+            iterator.throw(reason).then(noop, noop));
       }
       function abortIterable() {
         if (0 === streamTask.status) {
@@ -1357,7 +1357,7 @@
             : (erroredTask(request, streamTask, signal.reason),
               enqueueFlush(request));
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error);
+            iterator.throw(reason).then(noop, noop);
         }
       }
       var isIterator = iterable === iterator,
@@ -3904,6 +3904,22 @@
       if (hasOwnProperty.call(moduleExports, metadata[2]))
         return moduleExports[metadata[2]];
     }
+    function appendBackingEntry(backingStore, key, value) {
+      backingStore.data.append(key, value);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
+    function appendBackingFile(backingStore, key, value, filename) {
+      backingStore.data.append(key, value, filename);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
     function ReactPromise(status, value, reason) {
       this.status = status;
       this.value = value;
@@ -4216,7 +4232,7 @@
       var chunks = response._chunks,
         chunk = chunks.get(id);
       chunk ||
-        ((chunk = response._formData.get(response._prefix + id)),
+        ((chunk = response._formData.data.get(response._prefix + id)),
         (chunk =
           "string" === typeof chunk
             ? new ReactPromise(
@@ -4328,6 +4344,10 @@
         case "fulfilled":
           id = chunk.value;
           chunk = chunk.reason;
+          if (null !== chunk && "error" in chunk)
+            throw Error(
+              "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+            );
           for (
             var localLength = 0,
               rootArrayContexts = response._rootArrayContexts,
@@ -4419,23 +4439,20 @@
     function createMap(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Map initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Map.");
-      response = new Map(model);
       model.$$consumed = !0;
-      return response;
+      return new Map(model);
     }
     function createSet(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Set initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Set.");
-      response = new Set(model);
       model.$$consumed = !0;
-      return response;
+      return new Set(model);
     }
     function extractIterator(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Iterator initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Iterator.");
-      response = model[Symbol.iterator]();
       model.$$consumed = !0;
-      return response;
+      return model[Symbol.iterator]();
     }
     function createModel(response, model, parentObject, key) {
       return "then" === key && "function" === typeof model ? null : model;
@@ -4473,7 +4490,7 @@
           Error("Already initialized typed array.")
         )
       );
-      reference = response._formData.get(key).arrayBuffer();
+      reference = response._formData.data.get(key).arrayBuffer();
       if (initializingHandler) {
         var handler = initializingHandler;
         handler.deps++;
@@ -4517,7 +4534,7 @@
       var chunks = response._chunks;
       stream = new ReactPromise("fulfilled", stream, controller);
       chunks.set(id, stream);
-      response = response._formData.getAll(response._prefix + id);
+      response = response._formData.data.getAll(response._prefix + id);
       for (id = 0; id < response.length; id++)
         (chunks = response[id]),
           "string" === typeof chunks &&
@@ -4737,24 +4754,33 @@
               getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
             );
           case "K":
-            obj = value.slice(2);
-            obj = response._prefix + obj + "_";
-            key = new FormData();
-            response = response._formData;
-            arrayRoot = Array.from(response.keys());
-            for (value = 0; value < arrayRoot.length; value++)
-              if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+            key = value.slice(2);
+            obj = response._prefix + "_";
+            key = obj + key + "_";
+            arrayRoot = new FormData();
+            for (response = response._formData; ; ) {
+              value = response;
+              reference = value.keys;
+              null === reference &&
+                ((reference = value.keys = Array.from(value.data.keys())),
+                (value.keyPointer = 0));
+              value = reference[value.keyPointer];
+              if (void 0 === value) break;
+              if (value.startsWith(key)) {
+                reference = response.data.getAll(value);
                 for (
-                  var entries = response.getAll(reference),
-                    newKey = reference.slice(obj.length),
-                    j = 0;
-                  j < entries.length;
-                  j++
+                  var referencedFormDataKey = value.slice(key.length), i = 0;
+                  i < reference.length;
+                  i++
                 )
-                  key.append(newKey, entries[j]);
-                response.delete(reference);
-              }
-            return key;
+                  arrayRoot.append(referencedFormDataKey, reference[i]);
+                reference = response;
+                reference.data.delete(value);
+                reference.keyPointer++;
+              } else if (value.startsWith(obj)) break;
+              else response.keyPointer++;
+            }
+            return arrayRoot;
           case "i":
             return (
               (arrayRoot = value.slice(2)),
@@ -4921,10 +4947,11 @@
               arrayRoot
             );
           case "B":
-            return (
-              (obj = parseInt(value.slice(2), 16)),
-              response._formData.get(response._prefix + obj)
-            );
+            obj = parseInt(value.slice(2), 16);
+            response = response._formData.data.get(response._prefix + obj);
+            if (!(response instanceof Blob))
+              throw Error("Referenced Blob is not a Blob.");
+            return response;
           case "R":
             return parseReadableStream(response, value, void 0);
           case "r":
@@ -4962,7 +4989,7 @@
       return {
         _bundlerConfig: bundlerConfig,
         _prefix: formFieldPrefix,
-        _formData: backingFormData,
+        _formData: { data: backingFormData, keyPointer: -1, keys: null },
         _chunks: chunks,
         _closed: !1,
         _closedReason: null,
@@ -5021,44 +5048,44 @@
     }
     function decodeAction$1(body, serverManifest) {
       var formData = new FormData(),
-        action = null,
-        seenActions = new Set();
+        maybeActionKey = null;
       body.forEach(function (value, key) {
         key.startsWith("$ACTION_")
           ? key.startsWith("$ACTION_REF_")
-            ? seenActions.has(key) ||
-              (seenActions.add(key),
-              (value = "$ACTION_" + key.slice(12) + ":"),
-              (value = decodeBoundActionMetaData(body, serverManifest, value)),
-              (action = loadServerReference(serverManifest, value)))
-            : key.startsWith("$ACTION_ID_") &&
-              !seenActions.has(key) &&
-              (seenActions.add(key),
-              (value = key.slice(11)),
-              (action = loadServerReference(serverManifest, {
-                id: value,
-                bound: null
-              })))
+            ? (maybeActionKey = key)
+            : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
           : formData.append(key, value);
       });
-      return null === action
-        ? null
-        : action.then(function (fn) {
-            return fn.bind(null, formData);
-          });
+      if (null === maybeActionKey) return null;
+      var actionKey = maybeActionKey,
+        action = null;
+      if (actionKey.startsWith("$ACTION_REF_"))
+        (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+          (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+          (action = loadServerReference(serverManifest, body));
+      else if (actionKey.startsWith("$ACTION_ID_"))
+        (body = actionKey.slice(11)),
+          (action = loadServerReference(serverManifest, {
+            id: body,
+            bound: null
+          }));
+      else throw Error("Cannot handle action key. This is a bug in React.");
+      return action.then(function (fn) {
+        return fn.bind(null, formData);
+      });
     }
     function decodeFormState$1(actionResult, body, serverManifest) {
       var keyPath = body.get("$ACTION_KEY");
       if ("string" !== typeof keyPath) return Promise.resolve(null);
-      var metaData = null;
+      var actionKey = null;
       body.forEach(function (value, key) {
-        key.startsWith("$ACTION_REF_") &&
-          ((value = "$ACTION_" + key.slice(12) + ":"),
-          (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+        key.startsWith("$ACTION_REF_") && (actionKey = key);
       });
-      if (null === metaData) return Promise.resolve(null);
-      var referenceId = metaData.id;
-      return Promise.resolve(metaData.bound).then(function (bound) {
+      if (null === actionKey) return Promise.resolve(null);
+      var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+      body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+      var referenceId = body.id;
+      return Promise.resolve(body.bound).then(function (bound) {
         return null === bound
           ? null
           : [actionResult, keyPath, referenceId, bound.length - 1];
@@ -5641,7 +5668,7 @@
           entry = entry[1];
           if ("string" === typeof entry) {
             var response = response$jscomp$0;
-            response._formData.append(name, entry);
+            appendBackingEntry(response._formData, name, entry);
             var prefix = response._prefix;
             if (name.startsWith(prefix)) {
               var chunks = response._chunks;
@@ -5649,14 +5676,14 @@
               (chunks = chunks.get(name)) &&
                 resolveModelChunk(response, chunks, entry, name);
             }
-          } else response$jscomp$0._formData.append(name, entry);
+          } else appendBackingEntry(response$jscomp$0._formData, name, entry);
           iterator.next().then(progress, error);
         }
       }
       function error(reason) {
         reportGlobalError(response$jscomp$0, reason);
         "function" === typeof iterator.throw &&
-          iterator.throw(reason).then(error, error);
+          iterator.throw(reason).then(noop, noop);
       }
       var iterator = iterable[ASYNC_ITERATOR](),
         response$jscomp$0 = createResponse(
diff --git a/cjs/react-server-dom-rspack-server.edge.production.js b/cjs/react-server-dom-rspack-server.edge.production.js
index d057993..a8e0a42 100644
--- a/cjs/react-server-dom-rspack-server.edge.production.js
+++ b/cjs/react-server-dom-rspack-server.edge.production.js
@@ -932,7 +932,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
       erroredTask(request, streamTask, reason),
       enqueueFlush(request),
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error));
+        iterator.throw(reason).then(noop, noop));
   }
   function abortIterable() {
     if (0 === streamTask.status) {
@@ -946,7 +946,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
         : (erroredTask(request, streamTask, signal.reason),
           enqueueFlush(request));
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error);
+        iterator.throw(reason).then(noop, noop);
     }
   }
   iterable = iterable === iterator;
@@ -2144,6 +2144,22 @@ function requireModule(metadata) {
     return moduleExports[metadata[2]];
 }
 var RESPONSE_SYMBOL = Symbol();
+function appendBackingEntry(backingStore, key, value) {
+  backingStore.data.append(key, value);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
+function appendBackingFile(backingStore, key, value, filename) {
+  backingStore.data.append(key, value, filename);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
 function ReactPromise(status, value, reason) {
   this.status = status;
   this.value = value;
@@ -2515,7 +2531,7 @@ function getChunk(response, id) {
   var chunks = response._chunks,
     chunk = chunks.get(id);
   chunk ||
-    ((chunk = response._formData.get(response._prefix + id)),
+    ((chunk = response._formData.data.get(response._prefix + id)),
     (chunk =
       "string" === typeof chunk
         ? createResolvedModelChunk(response, chunk, id)
@@ -2616,6 +2632,10 @@ function getOutlinedModel(
     case "fulfilled":
       id = chunk.value;
       chunk = chunk.reason;
+      if (null !== chunk && "error" in chunk)
+        throw Error(
+          "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+        );
       for (
         var localLength = 0,
           rootArrayContexts = response._rootArrayContexts,
@@ -2698,23 +2718,20 @@ function getOutlinedModel(
 function createMap(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Map initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Map.");
-  response = new Map(model);
   model.$$consumed = !0;
-  return response;
+  return new Map(model);
 }
 function createSet(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Set initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Set.");
-  response = new Set(model);
   model.$$consumed = !0;
-  return response;
+  return new Set(model);
 }
 function extractIterator(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Iterator initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Iterator.");
-  response = model[Symbol.iterator]();
   model.$$consumed = !0;
-  return response;
+  return model[Symbol.iterator]();
 }
 function createModel(response, model, parentObject, key) {
   return "then" === key && "function" === typeof model ? null : model;
@@ -2752,7 +2769,7 @@ function parseTypedArray(
       Error("Already initialized typed array.")
     )
   );
-  reference = response._formData.get(key).arrayBuffer();
+  reference = response._formData.data.get(key).arrayBuffer();
   if (initializingHandler) {
     var handler = initializingHandler;
     handler.deps++;
@@ -2796,7 +2813,7 @@ function resolveStream(response, id, stream, controller) {
   var chunks = response._chunks;
   stream = new ReactPromise("fulfilled", stream, controller);
   chunks.set(id, stream);
-  response = response._formData.getAll(response._prefix + id);
+  response = response._formData.data.getAll(response._prefix + id);
   for (id = 0; id < response.length; id++)
     (chunks = response[id]),
       "string" === typeof chunks &&
@@ -3011,24 +3028,33 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
         );
       case "K":
-        obj = value.slice(2);
-        obj = response._prefix + obj + "_";
-        key = new FormData();
-        response = response._formData;
-        arrayRoot = Array.from(response.keys());
-        for (value = 0; value < arrayRoot.length; value++)
-          if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+        key = value.slice(2);
+        obj = response._prefix + "_";
+        key = obj + key + "_";
+        arrayRoot = new FormData();
+        for (response = response._formData; ; ) {
+          value = response;
+          reference = value.keys;
+          null === reference &&
+            ((reference = value.keys = Array.from(value.data.keys())),
+            (value.keyPointer = 0));
+          value = reference[value.keyPointer];
+          if (void 0 === value) break;
+          if (value.startsWith(key)) {
+            reference = response.data.getAll(value);
             for (
-              var entries = response.getAll(reference),
-                newKey = reference.slice(obj.length),
-                j = 0;
-              j < entries.length;
-              j++
+              var referencedFormDataKey = value.slice(key.length), i = 0;
+              i < reference.length;
+              i++
             )
-              key.append(newKey, entries[j]);
-            response.delete(reference);
-          }
-        return key;
+              arrayRoot.append(referencedFormDataKey, reference[i]);
+            reference = response;
+            reference.data.delete(value);
+            reference.keyPointer++;
+          } else if (value.startsWith(obj)) break;
+          else response.keyPointer++;
+        }
+        return arrayRoot;
       case "i":
         return (
           (arrayRoot = value.slice(2)),
@@ -3185,10 +3211,11 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           arrayRoot
         );
       case "B":
-        return (
-          (obj = parseInt(value.slice(2), 16)),
-          response._formData.get(response._prefix + obj)
-        );
+        obj = parseInt(value.slice(2), 16);
+        response = response._formData.data.get(response._prefix + obj);
+        if (!(response instanceof Blob))
+          throw Error("Referenced Blob is not a Blob.");
+        return response;
       case "R":
         return parseReadableStream(response, value, void 0);
       case "r":
@@ -3215,7 +3242,7 @@ function createResponse(bundlerConfig, formFieldPrefix, temporaryReferences) {
   return {
     _bundlerConfig: bundlerConfig,
     _prefix: formFieldPrefix,
-    _formData: backingFormData,
+    _formData: { data: backingFormData, keyPointer: -1, keys: null },
     _chunks: chunks,
     _closed: !1,
     _closedReason: null,
@@ -3272,44 +3299,44 @@ function decodeBoundActionMetaData(
 }
 function decodeAction$1(body, serverManifest) {
   var formData = new FormData(),
-    action = null,
-    seenActions = new Set();
+    maybeActionKey = null;
   body.forEach(function (value, key) {
     key.startsWith("$ACTION_")
       ? key.startsWith("$ACTION_REF_")
-        ? seenActions.has(key) ||
-          (seenActions.add(key),
-          (value = "$ACTION_" + key.slice(12) + ":"),
-          (value = decodeBoundActionMetaData(body, serverManifest, value)),
-          (action = loadServerReference(serverManifest, value)))
-        : key.startsWith("$ACTION_ID_") &&
-          !seenActions.has(key) &&
-          (seenActions.add(key),
-          (value = key.slice(11)),
-          (action = loadServerReference(serverManifest, {
-            id: value,
-            bound: null
-          })))
+        ? (maybeActionKey = key)
+        : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
       : formData.append(key, value);
   });
-  return null === action
-    ? null
-    : action.then(function (fn) {
-        return fn.bind(null, formData);
-      });
+  if (null === maybeActionKey) return null;
+  var actionKey = maybeActionKey,
+    action = null;
+  if (actionKey.startsWith("$ACTION_REF_"))
+    (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+      (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+      (action = loadServerReference(serverManifest, body));
+  else if (actionKey.startsWith("$ACTION_ID_"))
+    (body = actionKey.slice(11)),
+      (action = loadServerReference(serverManifest, {
+        id: body,
+        bound: null
+      }));
+  else throw Error("Cannot handle action key. This is a bug in React.");
+  return action.then(function (fn) {
+    return fn.bind(null, formData);
+  });
 }
 function decodeFormState$1(actionResult, body, serverManifest) {
   var keyPath = body.get("$ACTION_KEY");
   if ("string" !== typeof keyPath) return Promise.resolve(null);
-  var metaData = null;
+  var actionKey = null;
   body.forEach(function (value, key) {
-    key.startsWith("$ACTION_REF_") &&
-      ((value = "$ACTION_" + key.slice(12) + ":"),
-      (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+    key.startsWith("$ACTION_REF_") && (actionKey = key);
   });
-  if (null === metaData) return Promise.resolve(null);
-  var referenceId = metaData.id;
-  return Promise.resolve(metaData.bound).then(function (bound) {
+  if (null === actionKey) return Promise.resolve(null);
+  var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+  body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+  var referenceId = body.id;
+  return Promise.resolve(body.bound).then(function (bound) {
     return null === bound
       ? null
       : [actionResult, keyPath, referenceId, bound.length - 1];
@@ -3364,7 +3391,7 @@ exports.decodeReplyFromAsyncIterable = function (iterable, options) {
       var name = entry[0];
       entry = entry[1];
       if ("string" === typeof entry) {
-        response._formData.append(name, entry);
+        appendBackingEntry(response._formData, name, entry);
         var prefix = response._prefix;
         if (name.startsWith(prefix)) {
           var chunks = response._chunks;
@@ -3372,14 +3399,14 @@ exports.decodeReplyFromAsyncIterable = function (iterable, options) {
           (chunks = chunks.get(name)) &&
             resolveModelChunk(response, chunks, entry, name);
         }
-      } else response._formData.append(name, entry);
+      } else appendBackingEntry(response._formData, name, entry);
       iterator.next().then(progress, error);
     }
   }
   function error(reason) {
     reportGlobalError(response, reason);
     "function" === typeof iterator.throw &&
-      iterator.throw(reason).then(error, error);
+      iterator.throw(reason).then(noop, noop);
   }
   var iterator = iterable[ASYNC_ITERATOR](),
     response = createResponse(
diff --git a/cjs/react-server-dom-rspack-server.node.development.js b/cjs/react-server-dom-rspack-server.node.development.js
index 7b2b3f2..d178ad6 100644
--- a/cjs/react-server-dom-rspack-server.node.development.js
+++ b/cjs/react-server-dom-rspack-server.node.development.js
@@ -1440,7 +1440,7 @@
           erroredTask(request, streamTask, reason),
           enqueueFlush(request),
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error));
+            iterator.throw(reason).then(noop, noop));
       }
       function abortIterable() {
         if (0 === streamTask.status) {
@@ -1454,7 +1454,7 @@
             : (erroredTask(request, streamTask, signal.reason),
               enqueueFlush(request));
           "function" === typeof iterator.throw &&
-            iterator.throw(reason).then(error, error);
+            iterator.throw(reason).then(noop, noop);
         }
       }
       var isIterator = iterable === iterator,
@@ -4351,6 +4351,22 @@
       if (hasOwnProperty.call(moduleExports, metadata[2]))
         return moduleExports[metadata[2]];
     }
+    function appendBackingEntry(backingStore, key, value) {
+      backingStore.data.append(key, value);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
+    function appendBackingFile(backingStore, key, value, filename) {
+      backingStore.data.append(key, value, filename);
+      value = backingStore.keys;
+      null === value
+        ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+          (backingStore.keyPointer = 0))
+        : value.push(key);
+    }
     function ReactPromise(status, value, reason) {
       this.status = status;
       this.value = value;
@@ -4663,7 +4679,7 @@
       var chunks = response._chunks,
         chunk = chunks.get(id);
       chunk ||
-        ((chunk = response._formData.get(response._prefix + id)),
+        ((chunk = response._formData.data.get(response._prefix + id)),
         (chunk =
           "string" === typeof chunk
             ? new ReactPromise(
@@ -4775,6 +4791,10 @@
         case "fulfilled":
           id = chunk.value;
           chunk = chunk.reason;
+          if (null !== chunk && "error" in chunk)
+            throw Error(
+              "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+            );
           for (
             var localLength = 0,
               rootArrayContexts = response._rootArrayContexts,
@@ -4866,23 +4886,20 @@
     function createMap(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Map initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Map.");
-      response = new Map(model);
       model.$$consumed = !0;
-      return response;
+      return new Map(model);
     }
     function createSet(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Set initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Set.");
-      response = new Set(model);
       model.$$consumed = !0;
-      return response;
+      return new Set(model);
     }
     function extractIterator(response, model) {
       if (!isArrayImpl(model)) throw Error("Invalid Iterator initializer.");
       if (!0 === model.$$consumed) throw Error("Already initialized Iterator.");
-      response = model[Symbol.iterator]();
       model.$$consumed = !0;
-      return response;
+      return model[Symbol.iterator]();
     }
     function createModel(response, model, parentObject, key) {
       return "then" === key && "function" === typeof model ? null : model;
@@ -4920,7 +4937,7 @@
           Error("Already initialized typed array.")
         )
       );
-      reference = response._formData.get(key).arrayBuffer();
+      reference = response._formData.data.get(key).arrayBuffer();
       if (initializingHandler) {
         var handler = initializingHandler;
         handler.deps++;
@@ -4964,7 +4981,7 @@
       var chunks = response._chunks;
       stream = new ReactPromise("fulfilled", stream, controller);
       chunks.set(id, stream);
-      response = response._formData.getAll(response._prefix + id);
+      response = response._formData.data.getAll(response._prefix + id);
       for (id = 0; id < response.length; id++)
         (chunks = response[id]),
           "string" === typeof chunks &&
@@ -5184,24 +5201,33 @@
               getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
             );
           case "K":
-            obj = value.slice(2);
-            obj = response._prefix + obj + "_";
-            key = new FormData();
-            response = response._formData;
-            arrayRoot = Array.from(response.keys());
-            for (value = 0; value < arrayRoot.length; value++)
-              if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+            key = value.slice(2);
+            obj = response._prefix + "_";
+            key = obj + key + "_";
+            arrayRoot = new FormData();
+            for (response = response._formData; ; ) {
+              value = response;
+              reference = value.keys;
+              null === reference &&
+                ((reference = value.keys = Array.from(value.data.keys())),
+                (value.keyPointer = 0));
+              value = reference[value.keyPointer];
+              if (void 0 === value) break;
+              if (value.startsWith(key)) {
+                reference = response.data.getAll(value);
                 for (
-                  var entries = response.getAll(reference),
-                    newKey = reference.slice(obj.length),
-                    j = 0;
-                  j < entries.length;
-                  j++
+                  var referencedFormDataKey = value.slice(key.length), i = 0;
+                  i < reference.length;
+                  i++
                 )
-                  key.append(newKey, entries[j]);
-                response.delete(reference);
-              }
-            return key;
+                  arrayRoot.append(referencedFormDataKey, reference[i]);
+                reference = response;
+                reference.data.delete(value);
+                reference.keyPointer++;
+              } else if (value.startsWith(obj)) break;
+              else response.keyPointer++;
+            }
+            return arrayRoot;
           case "i":
             return (
               (arrayRoot = value.slice(2)),
@@ -5368,10 +5394,11 @@
               arrayRoot
             );
           case "B":
-            return (
-              (obj = parseInt(value.slice(2), 16)),
-              response._formData.get(response._prefix + obj)
-            );
+            obj = parseInt(value.slice(2), 16);
+            response = response._formData.data.get(response._prefix + obj);
+            if (!(response instanceof Blob))
+              throw Error("Referenced Blob is not a Blob.");
+            return response;
           case "R":
             return parseReadableStream(response, value, void 0);
           case "r":
@@ -5409,7 +5436,7 @@
       return {
         _bundlerConfig: bundlerConfig,
         _prefix: formFieldPrefix,
-        _formData: backingFormData,
+        _formData: { data: backingFormData, keyPointer: -1, keys: null },
         _chunks: chunks,
         _closed: !1,
         _closedReason: null,
@@ -5419,7 +5446,7 @@
       };
     }
     function resolveField(response, key, value) {
-      response._formData.append(key, value);
+      appendBackingEntry(response._formData, key, value);
       var prefix = response._prefix;
       if (key.startsWith(prefix)) {
         var chunks = response._chunks;
@@ -5478,44 +5505,44 @@
     }
     function decodeAction$1(body, serverManifest) {
       var formData = new FormData(),
-        action = null,
-        seenActions = new Set();
+        maybeActionKey = null;
       body.forEach(function (value, key) {
         key.startsWith("$ACTION_")
           ? key.startsWith("$ACTION_REF_")
-            ? seenActions.has(key) ||
-              (seenActions.add(key),
-              (value = "$ACTION_" + key.slice(12) + ":"),
-              (value = decodeBoundActionMetaData(body, serverManifest, value)),
-              (action = loadServerReference(serverManifest, value)))
-            : key.startsWith("$ACTION_ID_") &&
-              !seenActions.has(key) &&
-              (seenActions.add(key),
-              (value = key.slice(11)),
-              (action = loadServerReference(serverManifest, {
-                id: value,
-                bound: null
-              })))
+            ? (maybeActionKey = key)
+            : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
           : formData.append(key, value);
       });
-      return null === action
-        ? null
-        : action.then(function (fn) {
-            return fn.bind(null, formData);
-          });
+      if (null === maybeActionKey) return null;
+      var actionKey = maybeActionKey,
+        action = null;
+      if (actionKey.startsWith("$ACTION_REF_"))
+        (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+          (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+          (action = loadServerReference(serverManifest, body));
+      else if (actionKey.startsWith("$ACTION_ID_"))
+        (body = actionKey.slice(11)),
+          (action = loadServerReference(serverManifest, {
+            id: body,
+            bound: null
+          }));
+      else throw Error("Cannot handle action key. This is a bug in React.");
+      return action.then(function (fn) {
+        return fn.bind(null, formData);
+      });
     }
     function decodeFormState$1(actionResult, body, serverManifest) {
       var keyPath = body.get("$ACTION_KEY");
       if ("string" !== typeof keyPath) return Promise.resolve(null);
-      var metaData = null;
+      var actionKey = null;
       body.forEach(function (value, key) {
-        key.startsWith("$ACTION_REF_") &&
-          ((value = "$ACTION_" + key.slice(12) + ":"),
-          (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+        key.startsWith("$ACTION_REF_") && (actionKey = key);
       });
-      if (null === metaData) return Promise.resolve(null);
-      var referenceId = metaData.id;
-      return Promise.resolve(metaData.bound).then(function (bound) {
+      if (null === actionKey) return Promise.resolve(null);
+      var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+      body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+      var referenceId = body.id;
+      return Promise.resolve(body.bound).then(function (bound) {
         return null === bound
           ? null
           : [actionResult, keyPath, referenceId, bound.length - 1];
@@ -6320,14 +6347,14 @@
           _entry$value = _entry$value[1];
           "string" === typeof _entry$value
             ? resolveField(response, entry, _entry$value)
-            : response._formData.append(entry, _entry$value);
+            : appendBackingEntry(response._formData, entry, _entry$value);
           iterator.next().then(progress, error);
         }
       }
       function error(reason) {
         reportGlobalError(response, reason);
         "function" === typeof iterator.throw &&
-          iterator.throw(reason).then(error, error);
+          iterator.throw(reason).then(noop, noop);
       }
       var iterator = iterable[ASYNC_ITERATOR](),
         response = createResponse(
@@ -6375,7 +6402,7 @@
               var blob = new Blob(JSCompiler_object_inline_chunks_281, {
                 type: mimeType
               });
-              response._formData.append(name, blob, filename);
+              appendBackingFile(response._formData, name, blob, filename);
               pendingFiles--;
               if (0 === pendingFiles) {
                 for (blob = 0; blob < queuedFields.length; blob += 2)
diff --git a/cjs/react-server-dom-rspack-server.node.production.js b/cjs/react-server-dom-rspack-server.node.production.js
index 1e37108..ec814e9 100644
--- a/cjs/react-server-dom-rspack-server.node.production.js
+++ b/cjs/react-server-dom-rspack-server.node.production.js
@@ -955,7 +955,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
       erroredTask(request, streamTask, reason),
       enqueueFlush(request),
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error));
+        iterator.throw(reason).then(noop, noop));
   }
   function abortIterable() {
     if (0 === streamTask.status) {
@@ -969,7 +969,7 @@ function serializeAsyncIterable(request, task, iterable, iterator) {
         : (erroredTask(request, streamTask, signal.reason),
           enqueueFlush(request));
       "function" === typeof iterator.throw &&
-        iterator.throw(reason).then(error, error);
+        iterator.throw(reason).then(noop, noop);
     }
   }
   iterable = iterable === iterator;
@@ -2174,6 +2174,22 @@ function requireModule(metadata) {
     return moduleExports[metadata[2]];
 }
 var RESPONSE_SYMBOL = Symbol();
+function appendBackingEntry(backingStore, key, value) {
+  backingStore.data.append(key, value);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
+function appendBackingFile(backingStore, key, value, filename) {
+  backingStore.data.append(key, value, filename);
+  value = backingStore.keys;
+  null === value
+    ? ((backingStore.keys = Array.from(backingStore.data.keys())),
+      (backingStore.keyPointer = 0))
+    : value.push(key);
+}
 function ReactPromise(status, value, reason) {
   this.status = status;
   this.value = value;
@@ -2545,7 +2561,7 @@ function getChunk(response, id) {
   var chunks = response._chunks,
     chunk = chunks.get(id);
   chunk ||
-    ((chunk = response._formData.get(response._prefix + id)),
+    ((chunk = response._formData.data.get(response._prefix + id)),
     (chunk =
       "string" === typeof chunk
         ? createResolvedModelChunk(response, chunk, id)
@@ -2646,6 +2662,10 @@ function getOutlinedModel(
     case "fulfilled":
       id = chunk.value;
       chunk = chunk.reason;
+      if (null !== chunk && "error" in chunk)
+        throw Error(
+          "Expected an initialized chunk but got an initialized stream chunk instead. This payload may have been submitted by an older version of React."
+        );
       for (
         var localLength = 0,
           rootArrayContexts = response._rootArrayContexts,
@@ -2728,23 +2748,20 @@ function getOutlinedModel(
 function createMap(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Map initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Map.");
-  response = new Map(model);
   model.$$consumed = !0;
-  return response;
+  return new Map(model);
 }
 function createSet(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Set initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Set.");
-  response = new Set(model);
   model.$$consumed = !0;
-  return response;
+  return new Set(model);
 }
 function extractIterator(response, model) {
   if (!isArrayImpl(model)) throw Error("Invalid Iterator initializer.");
   if (!0 === model.$$consumed) throw Error("Already initialized Iterator.");
-  response = model[Symbol.iterator]();
   model.$$consumed = !0;
-  return response;
+  return model[Symbol.iterator]();
 }
 function createModel(response, model, parentObject, key) {
   return "then" === key && "function" === typeof model ? null : model;
@@ -2782,7 +2799,7 @@ function parseTypedArray(
       Error("Already initialized typed array.")
     )
   );
-  reference = response._formData.get(key).arrayBuffer();
+  reference = response._formData.data.get(key).arrayBuffer();
   if (initializingHandler) {
     var handler = initializingHandler;
     handler.deps++;
@@ -2826,7 +2843,7 @@ function resolveStream(response, id, stream, controller) {
   var chunks = response._chunks;
   stream = new ReactPromise("fulfilled", stream, controller);
   chunks.set(id, stream);
-  response = response._formData.getAll(response._prefix + id);
+  response = response._formData.data.getAll(response._prefix + id);
   for (id = 0; id < response.length; id++)
     (chunks = response[id]),
       "string" === typeof chunks &&
@@ -3041,24 +3058,33 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           getOutlinedModel(response, arrayRoot, obj, key, null, createSet)
         );
       case "K":
-        obj = value.slice(2);
-        obj = response._prefix + obj + "_";
-        key = new FormData();
-        response = response._formData;
-        arrayRoot = Array.from(response.keys());
-        for (value = 0; value < arrayRoot.length; value++)
-          if (((reference = arrayRoot[value]), reference.startsWith(obj))) {
+        key = value.slice(2);
+        obj = response._prefix + "_";
+        key = obj + key + "_";
+        arrayRoot = new FormData();
+        for (response = response._formData; ; ) {
+          value = response;
+          reference = value.keys;
+          null === reference &&
+            ((reference = value.keys = Array.from(value.data.keys())),
+            (value.keyPointer = 0));
+          value = reference[value.keyPointer];
+          if (void 0 === value) break;
+          if (value.startsWith(key)) {
+            reference = response.data.getAll(value);
             for (
-              var entries = response.getAll(reference),
-                newKey = reference.slice(obj.length),
-                j = 0;
-              j < entries.length;
-              j++
+              var referencedFormDataKey = value.slice(key.length), i = 0;
+              i < reference.length;
+              i++
             )
-              key.append(newKey, entries[j]);
-            response.delete(reference);
-          }
-        return key;
+              arrayRoot.append(referencedFormDataKey, reference[i]);
+            reference = response;
+            reference.data.delete(value);
+            reference.keyPointer++;
+          } else if (value.startsWith(obj)) break;
+          else response.keyPointer++;
+        }
+        return arrayRoot;
       case "i":
         return (
           (arrayRoot = value.slice(2)),
@@ -3215,10 +3241,11 @@ function parseModelString(response, obj, key, value, reference, arrayRoot) {
           arrayRoot
         );
       case "B":
-        return (
-          (obj = parseInt(value.slice(2), 16)),
-          response._formData.get(response._prefix + obj)
-        );
+        obj = parseInt(value.slice(2), 16);
+        response = response._formData.data.get(response._prefix + obj);
+        if (!(response instanceof Blob))
+          throw Error("Referenced Blob is not a Blob.");
+        return response;
       case "R":
         return parseReadableStream(response, value, void 0);
       case "r":
@@ -3245,7 +3272,7 @@ function createResponse(bundlerConfig, formFieldPrefix, temporaryReferences) {
   return {
     _bundlerConfig: bundlerConfig,
     _prefix: formFieldPrefix,
-    _formData: backingFormData,
+    _formData: { data: backingFormData, keyPointer: -1, keys: null },
     _chunks: chunks,
     _closed: !1,
     _closedReason: null,
@@ -3255,7 +3282,7 @@ function createResponse(bundlerConfig, formFieldPrefix, temporaryReferences) {
   };
 }
 function resolveField(response, key, value) {
-  response._formData.append(key, value);
+  appendBackingEntry(response._formData, key, value);
   var prefix = response._prefix;
   if (key.startsWith(prefix)) {
     var chunks = response._chunks;
@@ -3312,44 +3339,44 @@ function decodeBoundActionMetaData(
 }
 function decodeAction$1(body, serverManifest) {
   var formData = new FormData(),
-    action = null,
-    seenActions = new Set();
+    maybeActionKey = null;
   body.forEach(function (value, key) {
     key.startsWith("$ACTION_")
       ? key.startsWith("$ACTION_REF_")
-        ? seenActions.has(key) ||
-          (seenActions.add(key),
-          (value = "$ACTION_" + key.slice(12) + ":"),
-          (value = decodeBoundActionMetaData(body, serverManifest, value)),
-          (action = loadServerReference(serverManifest, value)))
-        : key.startsWith("$ACTION_ID_") &&
-          !seenActions.has(key) &&
-          (seenActions.add(key),
-          (value = key.slice(11)),
-          (action = loadServerReference(serverManifest, {
-            id: value,
-            bound: null
-          })))
+        ? (maybeActionKey = key)
+        : key.startsWith("$ACTION_ID_") && (maybeActionKey = key)
       : formData.append(key, value);
   });
-  return null === action
-    ? null
-    : action.then(function (fn) {
-        return fn.bind(null, formData);
-      });
+  if (null === maybeActionKey) return null;
+  var actionKey = maybeActionKey,
+    action = null;
+  if (actionKey.startsWith("$ACTION_REF_"))
+    (actionKey = "$ACTION_" + actionKey.slice(12) + ":"),
+      (body = decodeBoundActionMetaData(body, serverManifest, actionKey)),
+      (action = loadServerReference(serverManifest, body));
+  else if (actionKey.startsWith("$ACTION_ID_"))
+    (body = actionKey.slice(11)),
+      (action = loadServerReference(serverManifest, {
+        id: body,
+        bound: null
+      }));
+  else throw Error("Cannot handle action key. This is a bug in React.");
+  return action.then(function (fn) {
+    return fn.bind(null, formData);
+  });
 }
 function decodeFormState$1(actionResult, body, serverManifest) {
   var keyPath = body.get("$ACTION_KEY");
   if ("string" !== typeof keyPath) return Promise.resolve(null);
-  var metaData = null;
+  var actionKey = null;
   body.forEach(function (value, key) {
-    key.startsWith("$ACTION_REF_") &&
-      ((value = "$ACTION_" + key.slice(12) + ":"),
-      (metaData = decodeBoundActionMetaData(body, serverManifest, value)));
+    key.startsWith("$ACTION_REF_") && (actionKey = key);
   });
-  if (null === metaData) return Promise.resolve(null);
-  var referenceId = metaData.id;
-  return Promise.resolve(metaData.bound).then(function (bound) {
+  if (null === actionKey) return Promise.resolve(null);
+  var formFieldPrefix = "$ACTION_" + actionKey.slice(12) + ":";
+  body = decodeBoundActionMetaData(body, serverManifest, formFieldPrefix);
+  var referenceId = body.id;
+  return Promise.resolve(body.bound).then(function (bound) {
     return null === bound
       ? null
       : [actionResult, keyPath, referenceId, bound.length - 1];
@@ -3446,14 +3473,14 @@ exports.decodeReplyFromAsyncIterable = function (iterable, options) {
       _entry$value = _entry$value[1];
       "string" === typeof _entry$value
         ? resolveField(response, entry, _entry$value)
-        : response._formData.append(entry, _entry$value);
+        : appendBackingEntry(response._formData, entry, _entry$value);
       iterator.next().then(progress, error);
     }
   }
   function error(reason) {
     reportGlobalError(response, reason);
     "function" === typeof iterator.throw &&
-      iterator.throw(reason).then(error, error);
+      iterator.throw(reason).then(noop, noop);
   }
   var iterator = iterable[ASYNC_ITERATOR](),
     response = createResponse(
@@ -3501,7 +3528,7 @@ exports.decodeReplyFromBusboy = function (busboyStream, options) {
           var blob = new Blob(JSCompiler_object_inline_chunks_287, {
             type: mimeType
           });
-          response._formData.append(name, blob, filename);
+          appendBackingFile(response._formData, name, blob, filename);
           pendingFiles--;
           if (0 === pendingFiles) {
             for (blob = 0; blob < queuedFields.length; blob += 2)
