diff --git a/dist/index.js b/dist/index.js
index a917b94b3445931050449dcd263e9474636bf371..7b369012ba86bca5216ade81cdb5bbd8e4bef0cc 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -12971,6 +12971,7 @@ class IndependentSharedPlugin {
                     'TreeShakingSharedPlugin',
                     'IndependentSharedPlugin',
                     'ModuleFederationPlugin',
+                    'RspackModuleFederationPlugin',
                     'SharedUsedExportsOptimizerPlugin',
                     'HtmlWebpackPlugin',
                     'HtmlRspackPlugin',
