diff --git a/dist/cjs/ssr/types.d.cts b/dist/cjs/ssr/types.d.cts
index 9050041..e148444 100644
--- a/dist/cjs/ssr/types.d.cts
+++ b/dist/cjs/ssr/types.d.cts
@@ -1,8 +1,8 @@
 import { Manifest } from '../manifest.cjs';
 import { MakeRouteMatch } from '../Matches.cjs';
 export interface DehydratedMatch {
     i: MakeRouteMatch['id'];
-    b?: MakeRouteMatch['__beforeLoadContext'];
+    b?: Record<string, unknown>;
     l?: MakeRouteMatch['loaderData'];
     e?: MakeRouteMatch['error'];
     u: MakeRouteMatch['updatedAt'];
diff --git a/dist/esm/ssr/types.d.ts b/dist/esm/ssr/types.d.ts
index 9f57d89..f71eaac 100644
--- a/dist/esm/ssr/types.d.ts
+++ b/dist/esm/ssr/types.d.ts
@@ -1,8 +1,8 @@
 import { Manifest } from '../manifest.js';
 import { MakeRouteMatch } from '../Matches.js';
 export interface DehydratedMatch {
     i: MakeRouteMatch['id'];
-    b?: MakeRouteMatch['__beforeLoadContext'];
+    b?: Record<string, unknown>;
     l?: MakeRouteMatch['loaderData'];
     e?: MakeRouteMatch['error'];
     u: MakeRouteMatch['updatedAt'];
