diff --git a/dist/unstable/http/internal/preResponseHandler.d.ts b/dist/unstable/http/internal/preResponseHandler.d.ts
index efb146d88ed6fc653b97f421050db5e3b4bbb7de..37e505bf6e0cc7bb65ce6f9a05d170dcb14f902d 100644
--- a/dist/unstable/http/internal/preResponseHandler.d.ts
+++ b/dist/unstable/http/internal/preResponseHandler.d.ts
@@ -1,2 +1,6 @@
-export {};
-//# sourceMappingURL=preResponseHandler.d.ts.map
\ No newline at end of file
+import type * as Effect from "../../../Effect.ts";
+import type { HttpServerError } from "../HttpServerError.ts";
+import type { HttpServerRequest } from "../HttpServerRequest.ts";
+import type { HttpServerResponse } from "../HttpServerResponse.ts";
+export declare const requestPreResponseHandlers: WeakMap<object, (request: HttpServerRequest, response: HttpServerResponse) => Effect.Effect<HttpServerResponse, HttpServerError>>;
+export declare const appendPreResponseHandlerUnsafe: (request: HttpServerRequest, handler: (request: HttpServerRequest, response: HttpServerResponse) => Effect.Effect<HttpServerResponse, HttpServerError>) => void;
