diff --git a/dist/Schema.d.ts b/dist/Schema.d.ts
index 9547bd05cb7b91e5e5decc43b64c10a47a86186a..e58693c3742604ccb703045dedd259fca2c66b6e 100644
--- a/dist/Schema.d.ts
+++ b/dist/Schema.d.ts
@@ -10812,7 +10812,7 @@ export declare namespace Annotations {
          *
          * Reserved to internal use only.
          */
-        readonly "~sentinels"?: ReadonlyArray<SchemaAST.Sentinel> | undefined;
+        readonly "~sentinels"?: ReadonlyArray<unknown> | undefined;
     }
     /**
      * Annotations for filter schema nodes (created via `Schema.filter`). Extends
